src/tools/miri/src/alloc_addresses/mod.rs

```diff
@@ -166,7 +173,33 @@ trait EvalContextExtPriv<'tcx>: crate::MiriInterpCxExt<'tcx> {
                 assert!(!matches!(kind, AllocKind::Dead));
 
                 // This allocation does not have a base address yet, pick or reuse one.
-                let base_addr = if let Some((reuse_addr, clock)) = global_state.reuse.take_addr(
+                let base_addr = if ecx.machine.native_lib.is_some() {
+                    match kind {
+                        AllocKind::LiveData => {
+                            if ecx.tcx.try_get_global_alloc(alloc_id).is_some() {
+                                // For new global allocations, we always pre-allocate the memory to be able use the machine address directly.
+                                let prepared_bytes = MiriAllocBytes::zeroed(size, align)
+                                    .unwrap_or_else(|| {
+                                        panic!("Miri ran out of memory: cannot create allocation of {size:?} bytes")
+                                    });
+                                let addr = prepared_bytes.as_ptr().addr().try_into().unwrap();
```

Comment:
The `addr().try_into().unwrap()` happens in both `if` and `else`. Can you factor that out, i.e. something like
```rust
let ptr = if ecx.tcx.try_get_global_alloc(alloc_id).is_some() ...;
ptr.addr().try_into().unwrap()
```
And then we actually don't want to use `addr` here, we want to ensure the provenance is exposed:
```rust
let ptr = if ecx.tcx.try_get_global_alloc(alloc_id).is_some() ...;
// Ensure this pointer's provenance is exposed, so that it can be used by FFI code.
ptr.expose_provenance().try_into().unwrap()
```